misc/multistream-select/src/dialer_select.rs

```diff
@@ -177,13 +220,141 @@ where
+                        }
+                    }
+
+                    match protocol {
+                        None => {
+                            let msg = Message::Protocol(SIM_OPEN_ID);
+                            if let Err(err) = Pin::new(&mut io).start_send(msg) {
+                                return Poll::Ready(Err(From::from(err)));
+                            }
+
+                            let protocol = this.protocols.next().ok_or(NegotiationError::Failed)?;
+                            *this.state = SeqState::SendSimOpen {
+                                io,
+                                protocol: Some(protocol),
+                            };
+                        }
+                        Some(protocol) => {
+                            let p = Protocol::try_from(protocol.as_ref())?;
+                            if let Err(err) =
+                                Pin::new(&mut io).start_send(Message::Protocol(p.clone()))
+                            {
+                                return Poll::Ready(Err(From::from(err)));
+                            }
+                            log::debug!("Dialer: Proposed protocol: {}", p);
 
-                    // The dialer always sends the header and the first protocol
-                    // proposal in one go for efficiency.
-                    *this.state = SeqState::SendProtocol { io, protocol };
+                            *this.state = SeqState::FlushSimOpen { io, protocol }
+                        }
+                    }
                 }
 
+                SeqState::FlushSimOpen { mut io, protocol } => {
+                    match Pin::new(&mut io).poll_flush(cx)? {
+                        Poll::Ready(()) => *this.state = SeqState::AwaitSimOpen { io, protocol },
+                        Poll::Pending => {
+                            *this.state = SeqState::FlushSimOpen { io, protocol };
+                            return Poll::Pending;
+                        }
+                    }
+                }
+
+                SeqState::AwaitSimOpen { mut io, protocol } => {
+                    let msg = match Pin::new(&mut io).poll_next(cx)? {
+                        Poll::Ready(Some(msg)) => msg,
+                        Poll::Pending => {
+                            *this.state = SeqState::AwaitSimOpen { io, protocol };
+                            return Poll::Pending;
+                        }
+                        // Treat EOF error as [`NegotiationError::Failed`], not as
+                        // [`NegotiationError::ProtocolError`], allowing dropping or closing an I/O
+                        // stream as a permissible way to "gracefully" fail a negotiation.
+                        Poll::Ready(None) => return Poll::Ready(Err(NegotiationError::Failed)),
+                    };
+
+                    match msg {
+                        Message::Header(v) if v == HeaderLine::from(*this.version) => {
```

Comment:
This is equivalent I think? Given that `SimOpen` implies `V1`, I find it awkward to parse the version back into the headerline.

```suggestion
                        Message::Header(HeaderLine::V1) => {
```